src/webgpu/shader/validation/expression/call/builtin/dot.spec.ts

```diff
@@ -0,0 +1,189 @@
+const builtin = 'dot';
+export const description = `
+Validation tests for the ${builtin}() builtin.
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import { keysOf, objectsToRecord } from '../../../../../../common/util/data_tables.js';
+import { kValue } from '../../../../../util/constants.js';
+import {
+  Type,
+  kConvertableToFloatVectors,
+  scalarTypeOf,
+  ScalarType,
+} from '../../../../../util/conversion.js';
+import { QuantizeFunc, quantizeToF16, quantizeToF32 } from '../../../../../util/math.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+import {
+  fullRangeForType,
+  kConstantAndOverrideStages,
+  stageSupportsType,
+  validateConstOrOverrideBuiltinEval,
+} from './const_override_validation.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kValidArgumentTypes = objectsToRecord(kConvertableToFloatVectors);
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function quantizeFunctionForScalarType(type: ScalarType): QuantizeFunc<any> {
```

Comment:
Do we need to use `any` here? Could we not use `QuantizeFunc<number | bigint>` or something?

Reply:
Yeah, I struggled with this for a while. The problem is that if you say it's a `number | bigint` then every math operation you do throws up TypeScript errors saying that it can't perform math operations between those types. I do pass every incoming argument through the quantize function (which also serves as a cast when needed) to ensure that we're always using the same data types between args, but the TypeScript linter doesn't have a way of recognizing that.

`any` isn't desirable, but at least it keeps you from needing to do lots of redundant checks and casts for every step, and I figure it would be best if that type of type fudging was kept to a utility function.

(I should at least add comments about all this.)